File: 2022/go/3.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func char_to_priority(input int) int {
	lower_case := 96
    upper_case := 38
	result := 0

    if strings.ToLower(string(input)) != string(input) {
        result =  input - upper_case
    } else {
        result = input - lower_case
    }

	return result
}

func main() {

	input, err := os.ReadFile("../3.txt")
	if err != nil {
		fmt.Printf("Could not read the file due to this %s error \n", err)
	}
	sc := string(input)
	lines := strings.Split(sc, "\n")

	var score_part1 int = 0
	var score_part2 int = 0

	for i := 0; i < len(lines); i++ {

		n := len(lines[i]) / 2
		left := lines[i][0:n]
		right := lines[i][n:]
		set := make(map[int]bool) 

		for _, char := range left {	
			set[int(char)] = true
		}

		for k := range set {
			if strings.Contains(right, string(k)) {
				score_part1 += char_to_priority(k)
			}
		}
	}

	for i := 0; i < len(lines); i += 3 {

		first := lines[i]
		set := make(map[int]bool) 

		for _, char := range first {	
			set[int(char)] = true
		}

		for k := range set {
			if strings.Contains(lines[i + 1], string(k)) && strings.Contains(lines[i + 2], string(k)){
				score_part2 += char_to_priority(k)
			}
		}
	}
	
	fmt.Printf("Score 1: %v\n", score_part1)	
	fmt.Printf("Score 2: %v\n", score_part2)	
}

File: 2022/rust/3.rs
use std::fs;
use std::collections::HashSet;

fn char_to_priority(input : char) -> u32 {
    let lower_case : u32 = 96;
    let upper_case : u32 = 38;

    if input.is_uppercase() {
        input as u32 - upper_case
    }
    else {
        input as u32 - lower_case
    }
}


fn main() {
    let input = fs::read_to_string("../3.txt").unwrap();
    let lines : Vec<&str> = input.split("\n").collect();

    let mut score_part1 : u32 = 0;
    let mut score_part2 : u32 = 0;

    // Part 1
    for line in &lines {
        let (left, right) = line.trim().split_at(line.len() / 2);
        let half : HashSet<char> = left.chars().collect();

        for c in half {
            if right.contains(c) {
                score_part1 += char_to_priority(c);
            }
        }
    }

    // Part 2
    let mut i  = 0;
    let size = lines.len();
    while i < size - 1 {
        let first : HashSet<char> = lines[i].chars().collect();
        for c in first {
            if lines[i+1].contains(c) && lines[i+2].contains(c) {
                score_part2 += char_to_priority(c);
            }
        }
        i += 3;
    }
    
    println!("Score 1: {}", score_part1);
    println!("Score 2: {}", score_part2);
    
}